exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFunctionRegistry.java

```diff
@@ -64,62 +76,134 @@
       .put("CONVERT_FROM", Pair.of(2, 2))
       .put("FLATTEN", Pair.of(1, 1)).build();
 
+  /** Registers all functions present in Drill classpath on start-up. All functions will be marked as built-in.*/
   public DrillFunctionRegistry(ScanResult classpathScan) {
+    validate(BUILT_IN, classpathScan);
+    register(BUILT_IN, classpathScan, this.getClass().getClassLoader());
+    if (logger.isTraceEnabled()) {
+      StringBuilder allFunctions = new StringBuilder();
+      for (DrillFuncHolder method: registryHolder.getAllFunctionsWithHolders().values()) {
+        allFunctions.append(method.toString()).append("\n");
+      }
+      logger.trace("Registered functions: [\n{}]", allFunctions);
+    }
+  }
+
+  /**
+   * Validates all functions, present in jars.
+   * Will throw {@link FunctionValidationException} if:
+   * 1. Jar with the same name has been already registered.
+   * 2. Conflicting function with the similar signature is found.
+   * 3. Aggregating function is not deterministic.
+   *
+   * @return list of validated functions
+   */
+  public List<Func> validate(String jarName, ScanResult classpathScan) {
+    List<Func> functions = Lists.newArrayList();
     FunctionConverter converter = new FunctionConverter();
     List<AnnotatedClassDescriptor> providerClasses = classpathScan.getAnnotatedClasses();
 
-    // Hash map to prevent registering functions with exactly matching signatures
-    // key: Function Name + Input's Major Type
-    // value: Class name where function is implemented
-    //
-    final Map<String, String> functionSignatureMap = new HashMap<>();
+    if (registryHolder.containsJar(jarName)) {
+      throw new FunctionValidationException(String.format("Jar %s is already registered", jarName));
+    }
```

Comment:
Race condition. The check holds, then releases a lock. We proceed assuming the function is unique. But, another thread could be executing exactly in parallel, has also determined that the (same) function does not exist, and the two threads will now do the registration in parallel. Should this method be synchronized?